// react-titanium/lib/ReactTitaniumReconcileTransaction.js
'use strict';

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _reactLibCallbackQueue = require('react/lib/CallbackQueue');

var _reactLibCallbackQueue2 = _interopRequireDefault(_reactLibCallbackQueue);

var _reactLibPooledClass = require('react/lib/PooledClass');

var _reactLibPooledClass2 = _interopRequireDefault(_reactLibPooledClass);

var _reactLibTransaction = require('react/lib/Transaction');

var _reactLibTransaction2 = _interopRequireDefault(_reactLibTransaction);

var assign = _Object$assign;

var QUEUEING = {
  initialize: function initialize() {
    this.reactMountReady.reset();
  },
  close: function close() {
    this.reactMountReady.notifyAll();
  }
};

function ReactTitaniumReconcileTransaction() {
  this.reinitializeTransaction();
  this.reactMountReady = _reactLibCallbackQueue2['default'].getPooled(null);
}

var Mixin = {
  getTransactionWrappers: function getTransactionWrappers() {
    return [QUEUEING];
  },
  getReactMountReady: function getReactMountReady() {
    return this.reactMountReady;
  },
  destructor: function destructor() {
    _reactLibCallbackQueue2['default'].release(this.reactMountReady);
    this.reactMountReady = null;
  }
};

assign(ReactTitaniumReconcileTransaction.prototype, _reactLibTransaction2['default'].Mixin, Mixin);

_reactLibPooledClass2['default'].addPoolingTo(ReactTitaniumReconcileTransaction);

exports['default'] = ReactTitaniumReconcileTransaction;
module.exports = exports['default'];
